File: src/client/battle/report/EnvironmentShip.lua

-- 环境战船
-- @author alexdu
EnvironmentShip = class("EnvironmentShip")

function EnvironmentShip:ctor()
    self.captainType =  0

    --向性
    self.campPreference =  0

    --船只名称
    self.shipName =  ""

    --船只等级
    self.level =  1

    --性别
    self.gender =  0

    --普通技能模板
    self.commonSkillMould =  0

    --船只技能
    self.skillMould =  0

    --技能名称
    self.skillName =  ""

    --技能描述
    self.skillDescribe =  ""

    --战船类型，0为白色战船，1为绿色战船，2为蓝色战船，3为紫色战船，4为橙色战船
    self.shipType =  0

    --体力值
    self.power =  0

    --勇气值
    self.courage =  0

    --智力值
    self.intellect = 0

    --法防
    self.nimable =  0

    --防暴
    self.criticalResist =  0

    --闪避
    self.jink =  0

    --命中
    self.accuracy =  0

    --格挡
    self.retain =  0

    --破击
    self.retainBreak =  0

    --暴击系数
    self.critical =  0

    --拥有怒气值 0为拥有，1为没有
    self.haveDeadly =  0

    --体力成长参数
    self.growPower =  0

    --勇气成长参数
    self.growCourage =  0

    --智力成长参数
    self.growIntellect =  0

    --敏捷成长参数
    self.growNimable =  0

    --图片索引
    self.picIndex =  ""

    --半身像索引
    self.bustIndex =  -1

    --物理攻击特效
    self.physicsAttackEffect =  -1

    --屏幕攻击特效
    self.screenAttackEffect =  -1

    --物理攻击模式 0:原地攻击 1:移动到被攻击者船正前方攻击
    self.physicsAttackMode =  0

    --技能攻击模式 0:原地攻击 1:移动到被攻击者船正前方攻击
    self.skillAttackMode =  0

    --掉落卡片id
    self.rewardCard =  0

    --掉落卡片几率
    self.rewardCardRatio =  0

    --天赋模板
    self.talentMould =  ""

    --向性索引
    self.capacity =  0

    --生命加成百分比
    self.powerAdditionalPercent =  0

    --攻击加成百分比
    self.courageAdditionalPercent =  0

    --物理防御加成百分比
    self.intellectAdditionalPercent =  0

    --法防加成百分比
    self.nimableAdditionalPercent =  0

    --暴击加成百分比
    self.criticalAdditionalPercent =  0

    --抗暴加成百分比
    self.criticalResistAdditionalPercent =  0

    --闪避加成百分比
    self.evasionAdditionalPercent =  0

    --命中加成百分比
    self.accuracyAdditionalPercent =  0

    --挡格加成百分比
    self.retainAdditionalPercent =  0

    --破击加成百分比
    self.retainBreakAdditionalPercent =  0

    --物理免伤加成百分比
    self.physicalLessenDamagePercent =  0

    --法术免伤加成百分比
    self.skillLessenDamagePercent =  0

    --治疗率加成百分比
    self.curePercent =  0

    --被治疗率加成百分比
    self.byCurePercent =  0

    --最终伤害加成
    self.finalDamageAdditional =  0

    --最终免伤加成
    self.finalLessenDamageAdditional =  0

    --初始怒气增加
    self.initialSPIncrease =  0

    --天赋模板列表
    self.talentMouldList = {}

    --指向 ship_mould
    self.directing =  0

    --掉落道具
    self.dropProp =  0

    --掉落道具概率
    self.dropPropRatio =  0

    --掉落装备
    self.dropEquipment =  0

    --掉落装备概率
    self.dropEquipmentRatio =  0

    --根据影响值改变船只属性
    --@param influenceValue
    self._environmentShips =  nil
    --初始怒气槽
    self.zoariumSkillPoint = 0
end

function EnvironmentShip:addPropertyByInfluenceValue(influenceValue)
	local propertyTypeArray = zstring.split(influenceValue, "|")
	for key, property in pairs(propertyTypeArray) do
		local propertyType  = tonumber(zstring.split(property, IniUtil.comma)[1])
		local propertyValue = tonumber(zstring.split(property, IniUtil.comma)[2])
		-- _crint (propertyType, propertyValue)
		-- _crint (EquipmentMould.PROPERTY_TYPE_LIFE_ADDITIONAL)
		if (propertyType == EquipmentMould.PROPERTY_TYPE_LIFE_ADDITIONAL ) then
			-- _crint "Will add power"
			self.power = self.power + propertyValue
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_ATTACK_ADDITIONAL) then
			self.courage = self.courage + propertyValue				
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_PHYSIC_DEFENCE_ADDITIONAL ) then
			self.intellect = self.intellect + propertyValue
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_SKILL_DEFENCE_ADDITIONAL ) then
			self.nimable = self.nimable + propertyValue
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_LIFE_ADDITIONAL_PERCENTAGE ) then
			if(propertyValue < 1) then
				propertyValue = propertyValue * 100
			end
			self.powerAdditionalPercent = self.powerAdditionalPercent + propertyValue
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_ATTACK_ADDITIONAL_PERCENTAGE ) then
			if(propertyValue < 1)then
				propertyValue = propertyValue * 100
			end
			self.courageAdditionalPercent = self.courageAdditionalPercent + propertyValue
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_PHYSIC_DEFENCE_ADDITIONAL_PERCENTAGE ) then
			if(propertyValue < 1)then
				propertyValue = propertyValue * 100
			end
			self.intellectAdditionalPercent = self.intellectAdditionalPercent + propertyValue
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_SKILL_DEFENCE_ADDITIONAL_PERCENTAGE ) then
			if(propertyValue < 1)then
				propertyValue = propertyValue * 100
			end
			self.nimableAdditionalPercent = self.nimableAdditionalPercent + propertyValue
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_CRITICAL_ODDS ) then
			self.criticalAdditionalPercent = self.criticalAdditionalPercent + propertyValue
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_CRITICAL_RESIST_ODDS ) then
			self.criticalResistAdditionalPercent = self.criticalResistAdditionalPercent + propertyValue
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_RETAIN_ODDS ) then
			self.retainAdditionalPercent = self.retainAdditionalPercent + propertyValue
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_RETAIN_BREAK_ODDS ) then
			self.retainBreakAdditionalPercent = self.retainBreakAdditionalPercent + propertyValue
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_ACCURACY_ODDS ) then
			self.accuracyAdditionalPercent = self.accuracyAdditionalPercent + propertyValue
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_EVASION_ODDS ) then
			self.evasionAdditionalPercent = self.evasionAdditionalPercent + propertyValue
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_PHYSIC_LESSEN_DAMAGE_PERCENTAGE ) then
			self.physicalLessenDamagePercent = self.physicalLessenDamagePercent + (propertyValue / 100)
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_SKILL_LESSEN_DAMAGE_PERCENTAGE ) then
			self.skillLessenDamagePercent = self.skillLessenDamagePercent + (propertyValue / 100)
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_CURE_ODDS ) then
			self.curePercent = self.curePercent + (propertyValue / 100)
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_BY_CURE_ODDS ) then
			self.byCurePercent = self.byCurePercent + (propertyValue / 100)
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_FINAL_ADDITIONAL_DAMAGE ) then
			self.finalDamageAdditional = self.finalDamageAdditional + propertyValue
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_FINAL_LESSEN_DAMAGE ) then
			self.finalLessenDamageAdditional = self.finalLessenDamageAdditional + propertyValue
			
		elseif (propertyType == EquipmentMould.PROPERTY_TYPE_INITIAL_SP_INCREASE ) then
			self.initialSPIncrease = self.initialSPIncrease + propertyValue
        elseif (propertyType == EquipmentMould.PROPERTY_TYPE_FINAL_INIT_ZOMALSKILLPOINT) then
            self.zoariumSkillPoint = propertyValue
		else
		end
	end
end

function EnvironmentShip:calculateShipProperty()
	self:calculateShipPropertyByTalent()
	self.power = self.power * (100.0 + self.powerAdditionalPercent) / 100.0
	self.courage = self.courage * (100.0 + self.courageAdditionalPercent) / 100.0
	self.intellect = self.intellect * (100.0 + self.intellectAdditionalPercent) / 100.0
	self.nimable = self.nimable * (100.0 + self.nimableAdditionalPercent) / 100.0

	self.critical = self.critical + self.criticalAdditionalPercent
	self.jink = self.jink + self.evasionAdditionalPercent
end


-- 计算天赋影响船只能力
function EnvironmentShip:calculateShipPropertyByTalent()
	self.talentMouldList = {}
	----_crint ("self.talentMould = " .. self.talentMould)
	if( self.talentMould ~= "" and self.talentMould ~= "-1") then		
		local talentIdArray = zstring.split(self.talentMould, ",")
		--for (int i = 0; i < talentIdArray.length; i++) then
		for _, talentId in pairs(talentIdArray) do	
			--TalentMould talentMould = (TalentMould) Transformers.talentMouldDAOProxy().load(talentIdArray[i]);
			local talentMould = TalentMould:new()
			talentMould:init(dms.element(dms['talent_mould'], talentId) )
			if(talentMould.baseAdditional ~= "" and talentMould.baseAdditional ~= "-1") then					
				self:addPropertyByInfluenceValue(talentMould.baseAdditional);
			end
			table.insert(self.talentMouldList, talentMould)
		end
	end
end
	
--[[
local obj = EnvironmentShip:new()
obj:addPropertyByInfluenceValue("4, 0.6")
--_crint(obj.powerAdditionalPercent)
--]]

File: EgretProject/src/report/EnvironmentShip.ts
// environment_ship = {
// 	id = 1,									--	int(11)	Not null	主键，自增长
// 	captain_type = 2,					--	tinyint(4)		船长类型
// 	camp_preference = 3,			--	tinyint(4)		向性
// 	ship_name = 4,						--	varchar(16)		战船名称
// 	level = 5,								--	tinyint(4)		战船级别
// 	gender = 6,							--	tinyint(4)		性别(0:无 1:男 2:女)
// 	common_skill_mould = 7,		--	int(11)		普通技能模板
// 	skill_mould = 8,					--	int(11)		拥有技能，技能模版外键
// 	skill_name = 9,						--	varchar(16)		技能名称
// 	skill_describe = 10,				--	varchar(255)		技能描述
// 	ship_type = 11,						--	tinyint(4)		战船类型，0为白色战船，1为绿色战船，2为蓝色战船，3为紫色战船，4为橙色战船(卡牌低框)
// 	power = 12,							--	float		生命
// 	courage = 13,						--	float		攻击
// 	intellect = 14,						--	float		物防
// 	nimable = 15,						--	float		法防
// 	critical = 16,							--	float		暴击
// 	critical_resist = 17,				--	float		抗暴
// 	jink = 18,								--	float		闪避
// 	accuracy = 19,						--	float		命中
// 	retain = 20,							--	float		格挡
// 	retain_break = 21,					--	float		破击
// 	have_deadly = 22,					--			是否有怒气
// 	grow_power = 23,					--	int		体力成长参数，战船成长参数外键
// 	grow_courage = 24,				--	int		勇气成长参数，战船成长参数外键
// 	grow_intellect = 25,				--	int		智力成长参数，战船成长参数外键
// 	grow_nimable = 26,				--	int		敏捷成长参数，战船成长参数外键
// 	pic_index = 27,						--	varchar(255)		图片索引
// 	bust_index = 28,				--	Int(11)		半身像下标
// 	physics_attack_effect = 29,		--	int		物理攻击特效
// 	screen_attack_effect = 30,		--	int		屏幕攻击特效
// 	physics_attack_mode = 31,		--	tinyint(4)		物理攻击模式
// 	skill_attack_mode = 32,			--	tinyint(4)		技能攻击模式
// 	reward_card = 33,					--	int(11)		掉落卡片
// 	reward_card_ratio = 34,			--	int(11)		掉落几率
// 	talent_mould = 35,				-- varchar(255)		天赋模板
// 	capacity = 36,					-- int			战船标识(0:攻击型；1:防御型； 2：辅助型)
// 	directing = 37,					-- int 	战场模板指向
// 	drop_prop = 38,					--	int(11)		掉落道具
// 	drop_prop_ratio = 39,			--	int(11)		掉落概率
// 	drop_equipment = 40,			--	int(11)		掉落装备
// 	drop_equipment_ratio = 41,		--	int(11)		掉落概率
// 	sign_type = 42,					--	int(11)		BOSS标识
// 	sign_pic = 43,					--	int(11)		形象索引
// 	drop_items = 44 ,				--	vachar(255)		掉落库
// 	speed = 45,						--	int(11)		攻击
// 	common_attack_add_sp = 46,		--	int(11)		普攻回怒
// 	frist_drop_items = 47 ,			--	vachar(255)		首胜掉落库
// 	sound_index = 48,				--	int(11) 声音索引
// }

enum enumEnvironmentShip {
	id = 1,							
	captain_type = 2,				
	camp_preference = 3,			
	ship_name = 4,					
	level = 5,						
	gender = 6,						
	common_skill_mould = 7,			
	skill_mould = 8,				
	skill_name = 9,					
	skill_describe = 10,			
	ship_type = 11,				
	power = 12,						
	courage = 13,					
	intellect = 14,					
	nimable = 15,					
	critical = 16,					
	critical_resist = 17,			
	jink = 18,						
	accuracy = 19,					
	retain = 20,					
	retain_break = 21,				
	have_deadly = 22,				
	grow_power = 23,				
	grow_courage = 24,				
	grow_intellect = 25,			
	grow_nimable = 26,				
	pic_index = 27,					
	bust_index = 28,				
	physics_attack_effect = 29,		
	screen_attack_effect = 30,		
	physics_attack_mode = 31,		
	skill_attack_mode = 32,			
	reward_card = 33,				
	reward_card_ratio = 34,			
	talent_mould = 35,				
	capacity = 36,					
	directing = 37,					
	drop_prop = 38,					
	drop_prop_ratio = 39,			
	drop_equipment = 40,			
	drop_equipment_ratio = 41,		
	sign_type = 42,					
	sign_pic = 43,					
}

class EnvironmentShip {
	public constructor() {
	}

	id: any;						
	captainType: any;				
	campPreference: any;			
	shipName: any;				
	level: any;					
	gender: any;					
	commonSkillMould: any;		
	skillMould: any;				
	skillName: any;				
	skillDescribe: any;			
	shipType: any;			
	power: any;					
	courage: any;					
	intellect: any;					
	nimable: any;				
	critical: any;					
	criticalResist: any;			
	jink: any;						
	accuracy: any;					
	retain: any;					
	retainBreak: any;				
	haveDeadly: any;				
	growPower: any;				
	growCourage: any;				
	growIntellect: any;			
	growNimable: any;			
	picIndex: any;				
	bustIndex: any;			
	physicsAttackEffect: string;	
	screenAttackEffect: any;		
	physicsAttackMode: any;		
	skillAttackMode: any;			
	rewardCard: any;				
	rewardCardRatio: any;		
	talentMould: any;				
	capacity: any;					
	directing: any;					
	dropProp: any;				
	dropPropRatio: any;			
	dropEquipment: any;			
	dropEquipmentRatio: any;		
	signType: any;				
	signPic: any;

	public init(row) {
		this.id = row[0];						
		this.captainType = row[1];				
		this.campPreference = row[2];			
		this.shipName = row[3];				
		this.level = row[4];					
		this.gender = row[5];						
		this.commonSkillMould = row[6];			
		this.skillMould = row[7];				
		this.skillName = row[8];					
		this.skillDescribe = row[9];			
		this.shipType = row[10];				
		this.power = row[11];						
		this.courage = row[12];					
		this.intellect = row[13];					
		this.nimable = row[14];					
		this.critical = row[15];					
		this.criticalResist = row[16];			
		this.jink = row[17];						
		this.accuracy = row[18];					
		this.retain = row[19];					
		this.retainBreak = row[20];				
		this.haveDeadly = row[21];				
		this.growPower = row[22];				
		this.growCourage = row[23];				
		this.growIntellect = row[24];			
		this.growNimable = row[25];				
		this.picIndex = row[26];					
		this.bustIndex = row[27];				
		this.physicsAttackEffect = row[28];		
		this.screenAttackEffect = row[29];		
		this.physicsAttackMode = row[30];		
		this.skillAttackMode = row[31];			
		this.rewardCard = row[32];				
		this.rewardCardRatio = row[33];			
		this.talentMould = row[34];				
		this.capacity = row[35];					
		this.directing = row[36];					
		this.dropProp = row[37];					
		this.dropPropRatio = row[38];			
		this.dropEquipment = row[39];			
		this.dropEquipmentRatio = row[40];		
		this.signType = row[41];					
		this.signPic = row[42];
	}

}